automation/routine_evaluate_actions.go

```diff
@@ -0,0 +1,146 @@
+	// actions.
+	Annotations fields.Annotations
+
+	// Items lists item aggregations for items with known IDs.
+	Items []fields.ItemAggregation
+
+	// Calculations lists calculations to perform. A calculation can reference
+	// items or previous calculations.
+	Calculations []fields.Calculation
+
+	// SeriesIn filters which series keys (aliases) to return in the server
+	// result. If the value is nil, all series are returned.
+	SeriesIn []string
+}
+
+// EvaluateActions allows running a single evaluation and pass the result onto
+// a chained list of actions.
+type EvaluateActions struct {
+	// Evaluation contains the evaluations to perform.
+	Evaluation Evaluation
+
+	// TimeFunc is provided the current time, and should return a time range to
+	// evaluate. If not specified, a default window containing the last hour
+	// will be evaluated.
+	TimeFunc func(time.Time) (gte, lt time.Time)
+
+	// RollupBucket describe the rollup bucket to use for the evaluation. If not
+	// specified, a window rollup is used.
+	RollupBucket fields.CalendarDuration
+
+	// Actions describe a chain of functions that are run in order for each
+	// evaluation result. If an action modifies the evaluation result, this
+	// change is visible to the next action in the chain. If an action returns
+	// false, the chain is broken and no further actions are run for the given
+	// evaluation result.
+	//
+	// Actions are responsible for checking opts.DryRun, and for logging their
+	// own errors.
+	Actions []ActionFunc
+}
+
+func (e EvaluateActions) Do(ctx context.Context, cfg *Config) error {
+	logger := cfg.Logger()
+	client := cfg.Client()
+
+	var gte, lt time.Time
+
+	dataQuery := fields.Data().Where(fields.TimeRange(gte, lt))
+	if e.Evaluation.SeriesIn != nil {
+		dataQuery = dataQuery.Where(fields.SeriesIn(e.Evaluation.SeriesIn...))
+	}
+
+	selection, err := client.Clarify().
+		Evaluate(e.Evaluation.Items, e.Evaluation.Calculations, dataQuery).
+		Do(ctx)
+
+	if err != nil {
+		return err
+	}
+
+	result := EvaluateResult{
+		Annotations: maps.Clone(e.Evaluation.Annotations),
+		Data:        selection.Data,
+	}
+	logger.LogAttrs(
+		ctx, slog.LevelDebug, "Evaluation result",
+		slog.Any("annotations", result.Annotations),
+		slog.Any("data_frame", result.Data),
+	)
+	for _, action := range e.Actions {
+		if !action(ctx, cfg, &result) {
+			break
+		}
+	}
+	return nil
+}
+
+// ActionFunc describes a function that is run in response to an evaluation. The
+// return value indicate weather the next action in a chain of action should run
+// or not.
+//
+// Implementations are responsible for handling their own errors and for acting
+// according to the cfg.DryRun() and cfg.EarlyOut() configuration values.
+type ActionFunc func(ctx context.Context, cfg *Config, result *EvaluateResult) bool
+
+// ActionSeriesContains returns an action that return true if the named series
+// contains at least one instance of value.
```

Comment:
```suggestion
// ActionSeriesContains returns ActionFunc that checks if the named series
// contain at least one instance of value.
```

Reply:
Want to be explicit about the return value. I see standard lib often use _plain language_ rather than type names in doc strings. E.g. "returns a template" rather than "returns a Template". The type it returns is already given by the function _signature_ and we don't necessarily need to repeat it in the docs.

Reply:
Do you know why is it contain over contanis in this context?
Should we rename the function to ActionSeriesContain as well?

We have string.Contains in the standard lib.

Reply:
`Contans` is meant for singular form object, while `ActionSeries` is plural, so `ActionSeriesContain` makes more sense

Reply:
yes, but `template -> Template` makes exact match, while `action -> ActionFunc` doesn't.
> Want to be explicit about the return value. I see standard lib often use plain language rather than type names in doc strings. E.g. "returns a template" rather than "returns a Template". The type it returns is already given by the function signature and we don't necessarily need to repeat it in the docs.